### ISSUE TITLE
out variable is float

### ISSUE THREAD

**Issue by mkusner** [boluoweifenda/WAGE#8]
Maybe I'm misunderstanding but if I print both out and labels in the _loss function here: https://github.com/boluoweifenda/WAGE/blob/master/source/NN.py#L96

I get values that are floating point:
`This is labels: [[0 0 0]...]
This is out: [[-0.0114746094 0.12109375 0.0876464844]...]
`

I also see later that the error is cast to float32: https://github.com/boluoweifenda/WAGE/blob/master/source/NN.py#L113

But I thought all of the activations, weights, gradients, and errors are quantized to 2/8 bits?  I think I am likely misunderstanding something here. Please let me know if I am!

**Comment by boluoweifenda** [comment]
@mkusner 
Hi, thanks for your questions.
You can see that in the last fully connected layer, we have no activation, therefore, no Q(A) to quantize the output values
[https://github.com/boluoweifenda/WAGE/blob/master/source/NN.py#L88](url)
Then we can evaluate the model more precisely during test time:
`argmax(0.999,1) = 0`
or
`argmax(Q(0.999, 1)) = 1`
We have tested that the two methods produce similar result, and we choose the first one and leave out the Q(A).
Instead, we have Q(E) to ensure that the first error for the backpropagation process is quantized.
[https://github.com/boluoweifenda/WAGE/blob/master/source/NN.py#L91](url)

> I also see later that the error is cast to float32: 

I have to note that this 'error'  means the **error rate** of the classification results, not the **error tensors** mentioned in our paper that are used for bp calculation. Still, we keep this error rate at float32 to get precise evaluations.

**Comment by mkusner** [comment]
I see. This leads me to 2 questions:
1. But if you need the error rate to be float, then how is it possible to implement this on integer hardware? Or in fixed-point?
2. In general although you have quantization function in equation (3), am I correct in thinking you need to do the computation in floating point and then quantize?  I think this must be the case because if this was implemented in fixed point then you would never need the quantization equation (3).

**Comment by boluoweifenda** [comment]
@mkusner 
1. Calculating the error rate is just a  display of the training process. In hardware, you can directly omit this step, or you can count the total number of wrong samples, this not a big deal though.
2. Partially correct.
To be specific, the intermediate results are with higher precision, but not in floating point.
Eg, for a 8-bit input and 8-bit output multiplier, it needs 16 bits to store the intermediate results, and finally, cut out the high 8 bits as the output. (Most multipliers exactly do the same way) 
Our quantizations deal with the input and output bitwidths.

**Comment by mkusner** [comment]
Okay I think I understand: so let's say to multiply aW on line 3 of the algorithm we have a is some (b x h) matrix and W is some (h x r) matrix. If the activations are 8-bit and W is 2-bit then for each inner product of between 2 h-dimensional vectors we need precision of 10*h. If h is 100 we need the precision of 1000 bits. You're right this isn't floating point, but in your code your represent these numbers as floating point instead of 1000 bit numbers. Is that right?

**Comment by boluoweifenda** [comment]
@mkusner 
Maybe you misunderstand some points, let me give you an example:
y = W*x, x is a 256-dimension vector, W is a 256x256 matrix, then y is a 256-dimension vector.
So in the calculation of multiply-and-accumulate, yi (each point of y) is the accumulation of 256-dimension dot 256 dimension:
`yi= x1*Wi1 + x2*Wi2 + .... + x256*Wi256` 
So if W is INT8, x is INT8, we need at most 24 bit (the accumulation of 256 dot values with at most 16 bits).
 And in our experiment, in order to leverage the computation power of GPU, we just use a float32 to simulate the behavior of INT24.

**Comment by boluoweifenda** [comment]
@mkusner 
You can refer to the design of tpu-v1
[https://ieeexplore.ieee.org/abstract/document/8192463/](url)

**Comment by mkusner** [comment]
Can you explain why you just need 24 bits?

**Comment by boluoweifenda** [comment]
@mkusner 
`yi= x1*Wi1 + x2*Wi2 + .... + x256*Wi256 <= 2^8 * 2^8 + 2^8 * 2^8 + .....  = 256 * (2^8 * 2^8) = 2^24`
And the multiplier can be **reused** for each yi, if y has 256 dimensions, reuse for 256 times. Is that clear?
Forgive me but I don't understand why you confused about it.

**Comment by mkusner** [comment]
Ah yes sorry sorry I was thinking of multiplications. So you would have 10
+ log2(h) bits. For SVHN and CIFAR then you'd have 20-bits needed for the
fully connected layers. I see.

**Comment by boluoweifenda** [comment]
Thank you for the thoughtful discussion : )

**Comment by michaelklachko** [comment]
Wait, I think mkusner raised a valid point: `out` values are not quantized, so how are they implemented in 8 bit hardware? I'm a bit confused with your example:

> `argmax(0.999,1) = 0`
> or
> `argmax(Q(0.999, 1)) = 1`

Should it be argmax(0.999, 1) = 1, and argmax(Q(0.999), 1) = 0, with the second case being non-deterministic if Q(0.999) = 1? 

Regardless, if you only use 8 bit hardware, and the output distribution has two of very close values, both might be quantized to the same value, and the final output will be a random guess between them. 

I think you need to quantize model outputs in the code.

**Comment by boluoweifenda** [comment]
@michaelklachko 
Thanks for your reminding, I think it is a good point.
You are right, argmax(0.999,1) = 1 and argmax(Q(0.999, 1)) = 0 
> the output distribution has two of very close values, both might be quantized to the same value, and the final output will be a random guess between them

We have observed this in our previous experiments, but it happens rarely. Two methods produce similar accuracies and training curves, and we choose the first one and leave out the Q(A) in the paper to be more specific about the accuracy. Otherwise the order of labels will affect the inference results.
When it is implemented in INT8 hardware, the effect is negligible.
Since once the output has two very close and large values, e.g. (0.49 and 0.48), the model itself is already confused about the corresponding input and sensitive to any slight changes on pixel values (think about the adversarial sample). Then a random guess between them (with confidence coefficient) might be a better solution then a deterministic decision in real applications. This also explains why two methods have similar accuracies.
